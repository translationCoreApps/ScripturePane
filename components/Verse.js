import React from 'react';
import XRegExp from 'xregexp';
import usfmjs from 'usfm-js';
// helpers
import * as highlightHelpers from '../helpers/highlightHelpers';
import * as lexiconHelpers from '../helpers/lexiconHelpers';
// components
import WordDetails from './WordDetails';

class Verse extends React.Component {

  componentWillMount() {
    const {verseText} = this.props;
    if (verseText.constructor == Array) {
      this.props.verseText.forEach((word) => {
        const {strongs} = word
        const entryId = lexiconHelpers.lexiconEntryIdFromStrongs(strongs);
        const lexiconId = lexiconHelpers.lexiconIdFromStrongs(strongs);
        this.props.actions.loadLexiconEntry(lexiconId, entryId);
      });
    }
  }

  onClick(e, word) {
    let positionCoord = e.target;
    const PopoverTitle = <strong style={{fontSize: '1.2em'}}>{word.word}</strong>;
    let { showPopover } = this.props.actions;
    const wordDetails = <WordDetails {...this.props} word={word} />;
    showPopover(PopoverTitle, wordDetails, positionCoord);
  }

  verseArray(verseText = []) {
    let verseSpan = verseText.map( (word, index) => {
      return (
        <span style={{cursor: 'pointer'}} onClick={(e)=>this.onClick(e, word)} key={index}>
          {word.word + " "}
        </span>
      );
    });

    return verseSpan;
  }

  highlightQuoteInVerse(content, quote, occurrence) {
    let verseSpan = [];
    let regex = XRegExp('(?:^|\\PL)' + quote + '(?:$|\\PL)', 'g'); // use a regexp to capture the surrounding characters to use in rebuilding the verse
    let contentArray = XRegExp.split(content, regex); // split the verse into all of the pieces
    let aroundQuote = []; // store all of the characters surrounding the quote here
    XRegExp.forEach(content, regex, function (match, i) {
      aroundQuote[i] = match[0].split(quote); // store all of the characters surrounding the matches
    })
    let aroundQuoteIndex = occurrence - 1; // this ensures we can find the characters around the highlighted quote
    // for all of the content before the current quote
    // append the split quotes with their associated characters
    let beforeTextArray = contentArray.slice(0,occurrence).map( (text, i) => {
      if (i < occurrence-1) {   // if not the last one
        text = text + aroundQuote[i][0] + quote + aroundQuote[i][1]; // append the quote with the surrounding character
      }
      return text;
    })
    let beforeText = beforeTextArray.join(''); // join the pieces
    beforeText = beforeText + aroundQuote[aroundQuoteIndex][0]; // append the current quote's preceding char
    // for all of the content after the current quote
    // append the split quotes with their associated characters
    let afterTextArray = contentArray.slice(occurrence).map( (text, i) => {
      if (i < contentArray.length - occurrence - 1) { // if not the last one
        text = text + aroundQuote[i + occurrence][0] + quote + aroundQuote[i + occurrence][1]; // append the quote with the surrounding character
      }
      return text;
    })
    let afterText = afterTextArray.join('');  // join the pieces
    afterText = aroundQuote[aroundQuoteIndex][1] + afterText;  // prepend the current quote's preceding char
    verseSpan.push(
      <span key={1}>
        <span>{usfmjs.removeMarker(beforeText)}</span>
        <span style={{backgroundColor: "var(--highlight-color)"}}>
          {quote}
        </span>
        <span>{usfmjs.removeMarker(afterText)}</span>
      </span>
    );
    return verseSpan;
  }

  render() {
    let verseSpan = <span/>;
    let { verseText, chapter, verse, direction, bibleId, isCurrent} = this.props;
    if (typeof verseText === 'string') {
      let {quote, occurrence} = this.props.contextIdReducer.contextId;
      const isQuoteInVerse = highlightHelpers.isQuoteInVerse(verseText, quote);
      if (quote && verseText && isCurrent && bibleId === 'ulb' && !quote.includes("...") && isQuoteInVerse) {
        verseSpan = this.highlightQuoteInVerse(verseText, quote, occurrence);
      } else {
        verseSpan = <span>{usfmjs.removeMarker(verseText)}</span>;
      }
    } else {
      verseSpan = this.verseArray(verseText);
    }

    let chapterVerse = <strong>{chapter}:{verse} </strong>;
    if (direction === 'rtl') chapterVerse = <strong>{verse}:{chapter} </strong>;
    let divStyle = {direction: direction};

    return (
      <div style={divStyle}>
        {chapterVerse}
        {verseSpan}
      </div>
    );
  }
}

export default Verse;
