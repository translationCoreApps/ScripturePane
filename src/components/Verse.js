import React from 'react';
import PropTypes from 'prop-types';
// helpers
import * as lexiconHelpers from '../helpers/lexiconHelpers';
import {removeMarker} from '../helpers/UsfmHelpers';
import {isWord} from '../helpers/stringHelpers';
// components
import WordDetails from './WordDetails';
// constants
const PLACE_HOLDER_TEXT = '[WARNING: This Bible version does not include text for this reference.]';

class Verse extends React.Component {

  componentWillReceiveProps(nextProps) {
    if (nextProps.verseText && this.props.verseText !== nextProps.verseText) {
      if ((nextProps.verseText.constructor === Array) || (nextProps.verseText.verseObjects)) {
        const words = nextProps.actions.getWordListForVerse(nextProps.verseText);
        words.forEach((word) => {
          if (isWord(word)) {
            const {strong} = word;
            if (strong) {
              const entryId = lexiconHelpers.lexiconEntryIdFromStrongs(strong);
              const lexiconId = lexiconHelpers.lexiconIdFromStrongs(strong);
              nextProps.actions.loadLexiconEntry(lexiconId, entryId);
            }
          }
        });
      }
    }
  }

  onWordClick(e, word) {
    if (word && word.strong) {
      let positionCoord = e.target;
      const PopoverTitle = <strong style={{fontSize: '1.2em'}}>{word.word}</strong>;
      let {showPopover} = this.props.actions;
      const wordDetails = <WordDetails {...this.props} word={word}/>;
      showPopover(PopoverTitle, wordDetails, positionCoord);
    }
  }

  verseArray(verseText = []) {
    const { bibleId, contextIdReducer: { contextId } } = this.props;
    const words = this.props.actions.getWordListForVerse(verseText);
    let wordSpacing = '';
    const verseSpan = words.map((word, index) => {
      if (isWord(word)) {
        const padding = wordSpacing;
        wordSpacing = ' '; // spacing between words
        const text = (word.word || word.text);
        let isHighlightedWord = false;

        if (bibleId === 'ugnt') {
          isHighlightedWord = text === contextId.quote;
        } else if (bibleId === 'ulb') {
          isHighlightedWord = word.content === contextId.quote;
        }

        if (word.strong) { // if clickable
          return (
            <span
              key={index}
              onClick={(e) => this.onWordClick(e, word)}
              style={{ cursor: 'pointer', backgroundColor: isHighlightedWord ? "var(--highlight-color)" : "" }}
            >
              {padding + text}
            </span>
          );
        } else {
          return (
            <span
              key={index}
              style={{ backgroundColor: isHighlightedWord ? "var(--highlight-color)" : "" }}
            >
              {padding + text}
            </span>
          );
        }
      } else if (word.text) { // if not word, show punctuation, etc. but not clickable
        const lastChar = word.text.substr(word.text.length - 1);
        wordSpacing = ((lastChar === '"') || (lastChar === "'")) ? '' : ' '; // spacing before words
        return this.createTextSpan(index, word.text);
      }
    });

    return verseSpan;
  }

  createTextSpan(index, text) {
    return (
      <span key={index}>
        {text}
      </span>
    );
  }

  render() {
    let verseSpan = <span/>;
    let { verseText, chapter, verse, direction } = this.props;
    let verseIsPlaceHolder = false;

    if (!verseText) {
      verseText = PLACE_HOLDER_TEXT;
      verseIsPlaceHolder = true;
    }

    if (verseText && typeof verseText === 'string') { // if the verse content is string / text.
      verseSpan = <span>{removeMarker(verseText)}</span>;
    } else { // then the verse content is an array / verse objects.
      verseSpan = this.verseArray(verseText);
    }

    const chapterVerseContent = direction === 'rtl' ? `${verse}:${chapter} ` : `${chapter}:${verse} `;
    const chapterVerse = <strong>{chapterVerseContent}</strong>;
    let divStyle = { direction: direction };
    if (verseIsPlaceHolder) divStyle['fontStyle'] = 'italic';

    return (
      <div style={divStyle}>
        {chapterVerse}
        {verseSpan}
      </div>
    );
  }
}

Verse.propTypes = {
  actions: PropTypes.shape({
    setToolSettings: PropTypes.func.isRequired,
    getWordListForVerse: PropTypes.func.isRequired,
    loadLexiconEntry: PropTypes.func.isRequired,
    showPopover: PropTypes.func.isRequired
  }),
  verseText: PropTypes.oneOfType([
    PropTypes.string.isRequired,
    PropTypes.array.isRequired,
    PropTypes.object.isRequired
  ]),
  chapter: PropTypes.number.isRequired,
  verse: PropTypes.oneOfType([
    PropTypes.string.isRequired,
    PropTypes.number.isRequired
  ]),
  direction: PropTypes.string.isRequired,
  bibleId: PropTypes.string,
  isCurrent: PropTypes.bool.isRequired,
  contextIdReducer: PropTypes.object.isRequired
};

export default Verse;
export {PLACE_HOLDER_TEXT};
