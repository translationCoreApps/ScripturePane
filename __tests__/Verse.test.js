/* eslint-env jest */

import React from 'react';
import Verse, {PLACE_HOLDER_TEXT} from '../src/components/Verse';
import {mount} from 'enzyme';

// Tests for Verse React Component
describe('Test Verse component',()=>{
  test('Test when verse is null that the WARNING placeholder is displayed', () => {
    const props = {
      verseText: null,
      chapter: '1',
      verse: '1',
      contextIdReducer: {
        contextId: {}
      }
    };
    const expectedText = '1:1 '+PLACE_HOLDER_TEXT;
    const enzymeWrapper = mount(<Verse {...props} />);
    validateVerse(enzymeWrapper, expectedText);
  });

  test('Test when verse is not null that the verse is displayed', () => {
    const props = {
      verseText: 'Jesus wept',
      chapter: '11',
      verse: '35',
      contextIdReducer: {
        contextId: {}
      }
    };
    const expectedText = '11:35 Jesus wept';
    const enzymeWrapper = mount(<Verse {...props} />);
    validateVerse(enzymeWrapper, expectedText);
  });
});

function validateVerse(enzymeWrapper, expectedText) {
  const verseDiv = enzymeWrapper.find('div');
  expect(verseDiv.length).toEqual(1);
  expect(verseDiv.text()).toEqual(expectedText);
}
