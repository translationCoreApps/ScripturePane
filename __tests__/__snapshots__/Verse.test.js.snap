// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verse.js renders correctly 1`] = `
<div
  style={
    Object {
      "direction": "ltr",
    }
  }
>
  <strong>
    11:35 
  </strong>
  <span>
    Jesus wept
  </span>
</div>
`;
